### PR TITLE
Implement HttpProtocolException for HTTP/3

### LINKED ISSUE #70684: [API Proposal]: Exposing HTTP/2 and HTTP/3 protocol error codes
# Exposing HTTP/2 and HTTP/3 protocol error codes from `SocketsHttpHandler`

This proposal builds on our [`QuicException` proposal](https://github.com/dotnet/runtime/issues/70669).

## Background and Motivation

GRPC (and potentially other lower-level users) need a way to get the underlying HTTP/2 or HTTP/3 error codes in case a protocol error occurs. 

The error code should be observable not only when an error occurs while calling an `HttpClient` method (#43239), but also when invoking `Stream` API-s on the response's content read stream (#62228).

## Proposed design

- Define a new exception type `HttpProtocolException`, and embed it as `HttpRequestException.InnerException`
- Throw `ProtocolException` directly from `HttpResponse` content read streams

```C#
namespace System.Net.Http;

public sealed class HttpProtocolException : IOException
{
    public HttpProtocolException(HttpProtocolError errorCode, string message, Exception? innerException);

    public HttpProtocolError ErrorCode { get; }
}

// Map enum values to H2 and H3 error codes
// H3 error codes can be 62 bit long
public enum HttpProtocolError : long 
{
    // Camel-cased names taken directly from the HTTP/2 spec
    // https://datatracker.ietf.org/doc/html/rfc7540#section-7
    NoError = 0x0,
    ProtocolError,
    InternalError,
    FlowControlError,
    SettingsTimeout,
    StreamClosed,
    FrameSizeError,
    RefusedStream,
    Cancel,
    CompressionError,
    EnhanceYourCalm,
    InadequateSecurity,
    Http11Required,

    // Camel-cased names taken directly from the HTTP/3 spec
    // https://datatracker.ietf.org/doc/html/rfc9114/#section-8.1
    H3NoError = 0x100,
    H3GeneralProtocolError,
    H3InternalERror,
    H3StreamCreationError,
    H3ClosedCriticalStream,
    H3FrameUnexpected,
    H3FrameError,
    H3ExcessiveLoad,
    H3IdError,
    H3SettingsError,
    H3MissingSettings,
    H3RequestRejected,
    H3RequestCancelled,
    H3RequestIncomplete,
    H3MessageError,
    H3ConnectError,
    H3VersionFallback
}

```

**Edit:** Added a public constructor.

### Notes
- ~No public constructors, because we prefer to do the bare minimum to deliver a feature. This means that `WinHttpHandler` and user-made `HttpClientHandlers` won't be able to throw `HttpProtocolException` for now. We can consider public constructors later, if necessary.~
- We throw `ProtocolError` when a connection or stream is aborted, or when we detect a protocol violation ourselves
- `ErrorCode >= 256` means HTTP/3
- In case of HTTP/3 connection or stream errors, we are embedding `QuicException` as `ProtocolException.InnerException`
- We don't throw `HttpProtocolException` for transport-level errors

### Over HttpClient
```C#
using var client = new HttpClient();

try
{
    var response = await client.GetStringAsync(".");
}
catch (HttpRequestException ex) when (ex.InnerException is ProtocolException protocolException)
{
    Console.WriteLine("HTTP error code: " + protocolException.ProtocolErrorCode)
    if (protocolException.InnerException is QuicException quicException)
        Console.WriteLine("Underlying QUIC error: " + quicException.Message);
}
catch (HttpRequestException ex) when (ex.InnerException is AuthenticationException authenticationException) {
    // TLS authentication error. Can originate from QUIC, we map some errors to high-level .NET exceptions
}
```

### Over response stream

```C#
using var client = new HttpClient();
using var response = await client.GetAsync(".", HttpCompletionOption.ResponseHeadersRead);
using var responseStream = await response.Content.ReadAsStreamAsync();
using var memoryStream = new MemoryStream();

try {
    await responseStream.CopyToAsync(memoryStream);
}
catch (ProtocolException protocolException) {
    // HTTP(2|3) protocol error
}
catch (QuicException quicException) {
    // QUIC transport error
}
catch (IOException exception) when (ex.InnerException is SocketException socketException) {
    // TCP transport error
}
```

### Parallel independent exception types for HTTP/2 and HTTP/3

```C#
public class Http2ProtocolException : IOException
{
    public int ErrorCode { get; }
}

// Use System.Net.Quic.QuicException for HTTP/3
// public class QuicException : IOException { }
```

<details>
<summary>Usage</summary>

```C#
using var client = new HttpClient();

try
{
    var response = await client.GetStringAsync("foo.bar");
}
// HTTP/2
catch (HttpRequestException ex) when (ex.InnerException is Http2ProtocolException protocolException)
{
    Console.WriteLine(protocolException.ProtocolErrorCode)
}
// HTTP/3
catch (HttpRequestException ex) when (ex.InnerException is QuicException quicException)
{
    Console.WriteLine(quicException.ApplicationErrorCode);
}
```
</details>

### Do not define an enum, use untyped `long` error codes

```C#
public sealed class HttpProtocolException : IOException
{
    public long ErrorCode { get; }
}

```

## Risks
- `HttpProtocolError` enum: implementations may send unspecified error codes. If spec evolves in the future, we will have to introduce new error codes. In both cases field would contain a value outside the enum range, so users would have a workaround.
- Unless I'm missing something, this is not a breaking change, since `HttpProtocolException` is an `IOException`.

### PR DESCRIPTION
**Pull request by rzikm**
Follow-up on #71432.

Closes #70684.

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:1933]: Was this formatting alignment intentionally removed?

**Comment by ManickaP** [comment]
Please don't merge this before #71969 unless you;re 100% sure it will not create conflicts. I really need to get that PR in main.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:248]: Is it ok that we are ignoring both `_connection.AbortException` and `ex` now?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:1202]: Judging by `CreateHttp3StreamException` in `HttpProtocolException` it seems to me like `HttpProtocolException` may not always be connection-level. Is that right? This line does not seem as equivalent change then. Or am I missing something?

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs:38]: Should it be this instead?
```suggestion
            HttpProtocolException protocolEx = Assert.IsType<HttpProtocolException>(outerEx.InnerException);
            _output.WriteLine(protocolEx.Message);
```

Also, if I'm wrong and it's for diagnostics purposes in case it was not a HttpProtocolException, would it be better to include all the info about exceptions, even the top one? Meaning, should we log ToString instead of just Message

Comment (rzikm) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs:38]: I copied it from HttpClientHandlerTest.Http2 implementation, I can change it

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (rzikm) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:1202]: The HttpProtocolException thrown from stream is not on a code path which would affect this. In other words, only connection-level exceptions can be encountered here

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs
Comment (antonfirsov) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs]: This was somewhat accidental in my PR, we don't have to print anything, though it's informative for manual validation of `HttpProtocolException`'s message.

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs
Comment (antonfirsov) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs:382]:
```suggestion
```

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (antonfirsov) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:1201]: I'm wondering why are we throwing `IOException` below? Aren't these errors supposed to surface as `HttpProtocolException` from `HttpContent`'s `Http3ReadStream` (which seems to delegate to this method)?

**Review by antonfirsov — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
